man/v2syslog.3
.\" Copyright (C) 2022 VirtualSquare. Project Leader: Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" as published by the Free Software Foundation, either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.
.\"
.\" Automatically generated by Pandoc 3.1.11
.\"
.TH "V2SYSLOG" "3" "January 2024" "VirtualSquare" "Library Functions Manual"
.SH NAME
v2openlog, v2closelog, v2syslog, v2vsyslog, v2setlog, v2syslogx,
v2vsyslogx \- send messages to a local or networked logger
.SH SYNOPSIS
\f[CB]#include <v2syslog.h>\f[R]
.PP
\f[CB]void v2openlog(const char *\f[R] \f[I]ident\f[R]\f[CB], int\f[R]
\f[I]option,\f[R]\f[CB]int\f[R] \f[I]facility\f[R]\f[CB]);\f[R]
.PP
\f[CB]void v2closelog(void);\f[R]
.PP
\f[CB]void v2syslog(int\f[R] \f[I]priority\f[R]\f[CB], const char *\f[R]
\f[I]format\f[R]\f[CB], ...);\f[R]
.PP
\f[CB]void v2vsyslog(int\f[R]
\f[I]priority\f[R]\f[CB], const char *\f[R]
\f[I]format\f[R]\f[CB], va_list\f[R] \f[I]ap\f[R]\f[CB]);\f[R]
.PP
\f[CB]void v2setlog(struct ioth *\f[R]
\f[I]stack\f[R]\f[CB], struct v2syslog_server\f[R]
\f[I]server\f[R]\f[CB], const char *\f[R]
\f[I]hostname\f[R]\f[CB], const char *\f[R]
\f[I]procid\f[R]\f[CB]);\f[R]
.PP
\f[CB]void v2syslogx(int\f[R]
\f[I]priority\f[R]\f[CB], const char *\f[R]
\f[I]msg_id\f[R]\f[CB], const char *\f[R]
\f[I]struct_data\f[R]\f[CB], const char *\f[R]
\f[I]format\f[R]\f[CB], ...);\f[R]
.PP
\f[CB]void v2vsyslogx(int\f[R]
\f[I]priority\f[R]\f[CB], const char *\f[R]
\f[I]msg_id\f[R]\f[CB], const char *\f[R]
\f[I]struct_data\f[R]\f[CB], const char *\f[R]
\f[I]format\f[R]\f[CB], va_list\f[R] \f[I]ap\f[R]\f[CB]);\f[R]
.SH DESCRIPTION
\f[CB]v2syslog\f[R] is a library which enables daemon processes to route
their log messages to local or networked logging services.
.PP
It is an extension of \f[CB]syslog\f[R](3).
\f[CB]v2syslog\f[R] has been designed for Virtualsquare processes:
.IP \[bu] 2
Internet of Threads (IoTh) processes are network nodes, having their own
personality on the network (IP address, independent TCP\-IP stack).
The physical host (e.g.\ Linux box) they are currently running on is
merely incidental.
The local syslogd has no roles in this architecture.
.IP \[bu] 2
By VUOS users can run their own daemons in user space, using user
privileges.
All VUOS modules create specific execution environments for processes,
likewise if they were some sort of user level implemented namespaces:
neither root access nor specific capabilities are required.
Using the currently available syslog implementation users can run their
daemons but root access is required to read the log messsages
(e.g.\ read the /var/log/syslog file) or configure syslog to store or
process log messages elsewhere (e.g.\ by editing /etc/syslog.conf file)
.PP
This library provides the following functions:
.TP
\f[CB]v2openlog()\f[R]
\f[CB]v2openlog()\f[R] opens a connection to the system logger for a
program.
The arguments are the same of \f[CB]openlog\f[R](3).
The values that may be specified for \f[I]option\f[R] are described
below.
.TP
\f[CB]v2closelog()\f[R]
\f[CB]v2closelog()\f[R] closes the file descriptor being used to write
to the system logger.
The use of closelog() is optional.
.TP
\f[CB]v2syslog()\f[R], \f[CB]v2vsyslog()\f[R]
\f[CB]v2syslog()\f[R] generates and send a log message that can be
processed by \f[CB]syslogd\f[R](8) or \f[CB]v2syslogd\f[R](1) or any
other
logger using the protocols defined by RFC 3124 or RFC 5424.
The arguments are the same of \f[CB]syslog\f[R](3) or
\f[CB]vsyslog\f[R](3).
.TP
\f[CB]v2setlog()\f[R]
\f[CB]v2setlog()\f[R] defines the logging service to use.
\f[I]stack\f[R] is the Internet of Thread stack to use (the kernel
provided stack
is used if \f[I]stack\f[R] is NULL).
.RS
.PP
\f[I]server\f[R] defines the logging server address.
Some convenient macros are provided to set this argument:
.RE
.RS
.RS
.PP
\f[CB]v2syslog_UNIX(\f[R]\f[I]PATH\f[R]\f[CB])\f[R]: the deamon is
reachable at the UNIX socket \f[I]PATH\f[R].
.RE
.RE
.RS
.RS
.PP
\f[CB]v2syslog_INET(\f[R]\f[I]ADDR\f[R]\f[CB],\f[R]\f[I]PORT\f[R]\f[CB])\f[R]:
the daemon is available at the IPv4 address \f[I]ADDR\f[R], the port is
\f[I]PORT\f[R].
\f[I]ADDR\f[R] is a pointer to a \f[I]struct in_addr\f[R] (in network
byte order).
.RE
.RE
.RS
.RS
.PP
\f[CB]v2syslog_INET6(\f[R]\f[I]ADDR\f[R]\f[CB],\f[R]
\f[I]PORT\f[R]\f[CB])\f[R]: the daemon is available at the IPv6 address
\f[I]ADDR\f[R], the port is \f[I]PORT\f[R].
\f[I]ADDR\f[R] is a pointer to a \f[I]struct in6_addr\f[R] (in network
byte order).
.RE
.RE
.RS
.PP
\f[I]hostname\f[R] defines the hostname value included in log messages
(\f[I]nodename\f[R] returned by \f[CB]uname(2)\f[R] is used if
\f[I]hostname\f[R] is NULL)
.RE
.RS
.PP
\f[I]procid\f[R] defines the procid value for the log messages (the
output of \f[CB]getpid\f[R](2) or \f[CB]gettid\f[R](2) is used if this
argment is NULL).
.RE
.TP
\f[CB]v2syslogx()\f[R], \f[CB]v2vsyslogx()\f[R]
These functions extend \f[CB]v2syslog()\f[R] and \f[CB]v2vsyslog()\f[R]
respectively.
Two further arguments are provided:
\f[I]msg_id\f[R] define the message id (6.2.7 of RFC 5424),
\f[I]struct_data\f[R] define the structured\-date (6.3 of RFC 5424).
.TP
Values for \f[I]option\f[R]:
The \f[I]option\f[R] argument to \f[CB]v2openlog()\f[R] is a bit mask.
Its value is the bitwise\-or of the values defined for
\f[CB]openlog\f[R](3)
and the following values:
.RS
.PP
\f[CB]LOG_STREAM\f[R] try to send log messages using stream sockets (TCP
or UNIX\-STREAM)
.RE
.RS
.PP
\f[CB]LOG_STREAM_ONLY\f[R] send log messages using stream sockets (TCP
or UNIX\-STREAM) only
.RE
.RS
.PP
\f[CB]LOG_DGRAM_ONLY\f[R] send log messages using datagram sockets (UDP
or UNIX\-DGRAM) only
.RE
.RS
.PP
\f[CB]LOG_3164\f[R] use the protocol defined in RFC 3164
.RE
.RS
.PP
\f[CB]LOG_5424\f[R] use the protocol defined in RFC 5424
.RE
.RS
.PP
\f[CB]LOG_FRAMING_COUNT\f[R] use the octet counting (3.1.1 of RFC 6587)
.RE
.RS
.PP
\f[CB]LOG_USE_TID\f[R] send tid (thread id) instead of pid
.RE
.SH AUTHOR
VirtualSquare.
Project leader: Renzo Davoli


v2syslog.c
/*
 * v2syslog: syslog library for virtualsquare
 * Copyright (C) 2022  Renzo Davoli Virtualsquare University of Bologna
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program;
 * If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdarg.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <pthread.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/uio.h>
#include <sys/time.h>
#include <sys/utsname.h>
#include <netinet/in.h>
#include <syslog.h>
#include <ioth.h>

#include <v2syslog_const.h>
#include <v2syslog.h>

#define PID_STR_SIZE (((sizeof(pid_t) * 8  + 2) / 3) + 1)

static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
extern char *__progname;
static const char *syslog_ident;
static int syslog_option;
static int syslog_facility = LOG_USER;
static int syslog_fd = -1;
static int syslog_type = SOCK_DGRAM;
static char *syslog_procid;
static char *syslog_hostname;
static struct ioth *syslog_stack;
static union {
	struct sockaddr addr;
	struct sockaddr_un un;
	struct sockaddr_in in;
	struct sockaddr_in6 in6;
} syslog_sock;

static socklen_t addrlen(void) {
	switch(syslog_sock.addr.sa_family) {
		case AF_UNIX:
		default: return sizeof(syslog_sock.un);
		case AF_INET: return sizeof(syslog_sock.in);
		case AF_INET6: return sizeof(syslog_sock.in6);
	}
}

static struct sockaddr *addr(void) {
	return &syslog_sock.addr;
}

static int family(void) {
	if (syslog_sock.addr.sa_family == AF_UNSPEC) {
		syslog_sock.un.sun_family = AF_UNIX;
		snprintf(syslog_sock.un.sun_path, sizeof(syslog_sock.un.sun_path), "%s", LOG_DEFAULT_PATH);
	}
	return syslog_sock.addr.sa_family;
}

static int change_syslog_type(void) {
	switch (syslog_type) {
		case SOCK_DGRAM:
			if (syslog_option & LOG_ONLY) return -1;
			syslog_type = SOCK_STREAM;
			break;
		case SOCK_STREAM:
			if (syslog_option & LOG_ONLY) return -1;
			syslog_type = SOCK_DGRAM;
			break;
	}
	return 0;
}

/* open and connect a socket */
static int __openlog(void) {
	syslog_fd = ioth_msocket(syslog_stack, family(), syslog_type | SOCK_CLOEXEC, 0);
	if (syslog_fd == -1)
		return -1;
	if (ioth_connect(syslog_fd, addr(), addrlen()) == -1) {
		int saved_errno = errno;
		ioth_close(syslog_fd);
		syslog_fd = -1;
		return errno = saved_errno, -1;
	}
	return 0;
}

#define strdup_update(X, Y) do { \
	if ((X) != NULL) free(X); \
	(X) = ((Y) == NULL || (Y)[0] == 0) ? NULL : strdup(Y); \
} while (0);

void v2setlog(struct ioth *stack,struct v2syslog_server server,
    const char *hostname, const char *procid) {
	pthread_mutex_lock(&mutex);
	syslog_stack = stack;
	if (syslog_fd != -1) ioth_close(syslog_fd);
	syslog_fd = -1;

	if (server.port == 0) server.port = LOG_DEFAULT_PORT;
	memset(&syslog_sock, 0, sizeof(syslog_sock));
	switch(server.af) {
		case AF_UNIX:
		default:
			syslog_sock.un.sun_family = AF_UNIX;
			if (server.addr == NULL) server.addr = LOG_DEFAULT_PATH;
			snprintf(syslog_sock.un.sun_path, sizeof(syslog_sock.un.sun_path), "%s", server.straddr);
			if (syslog_sock.un.sun_path[0] == '~') {
				char *home = secure_getenv("HOME");
				if (home == NULL) home = "/";
				if (syslog_sock.un.sun_path[1] == 0)
					snprintf(syslog_sock.un.sun_path, sizeof(syslog_sock.un.sun_path),
							"%s" USER_LOG_DEFAULT_PATH, home);
				else
					snprintf(syslog_sock.un.sun_path, sizeof(syslog_sock.un.sun_path),
							"%s/%s", home, server.straddr);
			}
			break;
		case AF_INET:
			syslog_sock.in.sin_family = AF_INET;
			syslog_sock.in.sin_addr = *((struct in_addr *) server.addr);
			syslog_sock.in.sin_port = htons(server.port);
			break;
		case AF_INET6:
			syslog_sock.in6.sin6_family = AF_INET6;
			syslog_sock.in6.sin6_addr = *((struct in6_addr *) server.addr);
			syslog_sock.in6.sin6_port = htons(server.port);
			break;
	}
	strdup_update(syslog_hostname, hostname);
	strdup_update(syslog_procid, procid);
	pthread_mutex_unlock(&mutex);
}

void v2openlog(const char *ident, int option, int facility) {
	pthread_mutex_lock(&mutex);
	syslog_ident = (ident != NULL) ? ident : __progname;
	syslog_option = option;
	facility &= LOG_FACMASK;
	syslog_facility = (facility > 0 && LOG_FAC(facility) < LOG_NFACILITIES) ? facility : LOG_USER;
	syslog_type = (syslog_option & LOG_STREAM) ? SOCK_STREAM : SOCK_DGRAM;
	if (syslog_hostname == NULL) {
		struct utsname uts;
		uname(&uts);
		syslog_hostname = strdup(uts.nodename);
	}
	if (option & LOG_NDELAY) {
		if (__openlog() < 0 && (change_syslog_type() < 0 || __openlog() < 0)) {
			syslog_fd = -1;
			if (syslog_option & LOG_CONS)
				fprintf(stderr, "log server is unreachable\n");
		}
	}
	pthread_mutex_unlock(&mutex);
}

void v2closelog(void) {
	pthread_mutex_lock(&mutex);
	if (syslog_fd != -1) ioth_close(syslog_fd);
	syslog_fd = -1;
	pthread_mutex_unlock(&mutex);
}

static const char *get_syslog_pid(char *buf) {
	if (syslog_procid != NULL)
		return syslog_procid;
	snprintf(buf, PID_STR_SIZE, "%d",
#ifdef _GNU_SOURCE
			(syslog_option & LOG_USE_TID) ? gettid() : getpid()
#else
			getpid()
#endif
			);
	return buf;
}

/* avoid locale translation of monthes, setlocale would have affected the entire process */
static char *no_locale_months[] = {
	"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
};

static void header_3164(FILE *f, int priority) {
	fprintf (f, "<%d>", syslog_facility + priority);
	char timestr[13];
	time_t now;
	struct tm tmbuf;
	time(&now);
	if (syslog_ident == NULL) syslog_ident = __progname;
	localtime_r(&now, &tmbuf);
	strftime(timestr, 13, " %e %T", localtime_r(&now, &tmbuf));
	fprintf(f, "%s%s ", no_locale_months[tmbuf.tm_mon], timestr);
	if (syslog_hostname != NULL && (syslog_option & LOG_3164)) // 3164 requires id
		fprintf(f, "%s ", syslog_hostname);
	fprintf(f, "%s", syslog_ident);
	char pid_str_buf[PID_STR_SIZE];
	if (syslog_option & LOG_PID)
		fprintf(f, "[%s]: ", get_syslog_pid(pid_str_buf));
	else
		fprintf(f, ": ");
}

static void header_5424(FILE *f, int priority, const char *msg_id, const char *struct_data) {
	fprintf (f, "<%d>1 ", syslog_facility + priority);
	char timestr[21];
	struct timeval now;
	gettimeofday(&now, NULL);
	struct tm tmbuf;
	if (syslog_ident == NULL) syslog_ident = __progname;
	strftime(timestr, 21, "%Y-%m-%dT%H:%M:%S.", gmtime_r(&now.tv_sec, &tmbuf));
	char pid_str_buf[PID_STR_SIZE];
	fprintf(f, "%s%06ldZ %s %s %s %s %s ", timestr, now.tv_usec,
			(syslog_hostname != NULL) ? syslog_hostname : "-",
			syslog_ident,
			(syslog_option & LOG_PID) ? get_syslog_pid(pid_str_buf) : "-",
			(msg_id != NULL) ? msg_id : "-",
			(struct_data != NULL) ? struct_data : "-");
}

static int v2syslog_send(char *buf, size_t bufsize) {
	if (syslog_type == SOCK_DGRAM)
		return ioth_send(syslog_fd, buf, bufsize, 0);
	else {
		if (syslog_option & LOG_FRAMING_COUNT) {
			size_t lenlen = snprintf(NULL, 0, "%ld ", bufsize);
			char lenbuf[lenlen];
			snprintf(lenbuf, lenlen, "%ld ", bufsize);
			struct iovec iov[] = {{lenbuf, lenlen}, {buf, bufsize}};
			return ioth_writev(syslog_fd, iov, 2);
		} else
			return ioth_send(syslog_fd, buf, bufsize + 1, 0);
	}
}

static int v2syslog_stderr(char *buf, size_t bufsize) {
	char nl[]="\n";
	struct iovec iov[] = {{buf, bufsize}, {nl, 1}};
	return writev(STDERR_FILENO, iov, 2);
}

void v2vsyslogx(int priority, const char *msg_id, const char *struct_data,
		const char *format, va_list ap) {
	char *buf = 0;
	size_t bufsize = 0;
	int failure = 0;
	FILE *f = open_memstream(&buf, &bufsize);
	pthread_mutex_lock(&mutex);
	priority &= LOG_PRIMASK;
	if (syslog_option & LOG_5424)
		header_5424(f, priority, msg_id, struct_data);
	else
		header_3164(f, priority);
	long msg_offset = ftell(f);
	vfprintf(f, format, ap);
	fclose(f);
	if (syslog_fd == -1) {
		family(); // set to AF_UNIX if it is AF_UNSPEC
		if (__openlog() < 0 && (change_syslog_type() < 0 || __openlog() < 0))
			syslog_fd = -1, failure = 1;
	}
	if (syslog_fd != -1) {
		if (v2syslog_send(buf, bufsize) < 0) {
			ioth_close(syslog_fd);
			if (__openlog() < 0 && (change_syslog_type() < 0 || __openlog() < 0))
				syslog_fd = -1, failure = 1;
			else
				v2syslog_send(buf, bufsize);
		}
	}
	if (syslog_option & LOG_PERROR ||
			((syslog_option & LOG_CONS) && failure))
		v2syslog_stderr(buf + msg_offset, bufsize - msg_offset);
	pthread_mutex_unlock(&mutex);
}

void v2vsyslog(int priority, const char *format, va_list ap) {
	v2vsyslogx(priority, NULL, NULL, format, ap);
}

void v2syslogx(int priority, const char *msg_id, const char *struct_data, const char *format, ...) {
	va_list ap;
	va_start(ap, format);
	v2vsyslogx(priority, msg_id, struct_data, format, ap);
	va_end(ap);
}

void v2syslog(int priority, const char *format, ...) {
	va_list ap;
	va_start(ap, format);
	v2vsyslog(priority, format, ap);
	va_end(ap);
}
